### PR TITLE
[asm igemm][gfx90a][bf16]Asm igemm nhwc native bf16 kernel on gfx90a

### PR DESCRIPTION
**Pull request by shaojiewang**
NHWC: native bf16 asm igemm kernel on gfx90a.
- [x] fwd
- [x] bwd
- [x] wrw

ssd: [ssd-bf16](https://amdcloud-my.sharepoint.com/:x:/r/personal/carhuang_amd_com/_layouts/15/Doc.aspx?sourcedoc=%7B4A0FEC77-7042-4FEA-8BE4-F276E1200A16%7D&file=miopen-perf-ssd.xlsx&wdOrigin=OFFICECOM-WEB.START.REC&ct=1634825356337&action=default&mobileredirect=true)

### CODE REVIEW

**Comment by atamazov** [comment]
@shaojiewang In which release this needs to be included?

**Comment by junliume** [comment]
@shaojiewang How many tests can we enable on "Bf16 Hip Install All gfx90a" stage now?

File: src/include/miopen/solver.hpp [hunk 1]
```diff
@@ -2672,7 +2672,7 @@ struct PerformanceConfigAsmImplicitGemmGTC : Serializable<PerformanceConfigAsmIm
     template <class Self, class F>
     static void Visit(Self&& self, F f)
     {
-        std::string prec_string = self.precision == miopenFloat ? "fp32" : "fp16";
+        std::string prec_string = self.precision == miopenFloat ? "fp32" : (self.precision == miopenHalf ? "fp16" : "bf16");
```
[the remark this answers is not in the document]
Comment (carlushuang) [reply src/include/miopen/solver.hpp]: @atamazov [conv_asm_implicit_gemm_gtc_fwd_nhwc.cpp#L47](https://github.com/ROCmSoftwarePlatform/MIOpen/blob/develop/src/solver/conv_asm_implicit_gemm_gtc_fwd_nhwc.cpp#L47), here is a list of pre-generated configs. Actually fp16, fp32 may have different parameters, due to different optimization technique. So it is hard not to store the precision, and unify them with the same.

[the remark this answers is not in the document]
Comment (atamazov) [reply src/include/miopen/solver.hpp]: @carlushuang Datatype, direction and layout are parameters that belong to the problem. They do not belong to the tuning configs. IIRC you are your some kind of generator to produce kernels and pre-generated configs. If the generator produces more initializers than actually required by PerformanceConfig, then something is wrong with the design.

In this specific case I would recommend changing the generator so it produces separate table per each combination of datatype/direction/layout. Current design simply wastes memory.

Or you can use type other than `PerformanceConfigAsmImplicitGemmGTCFwdXdlopsNHWC` for the vector returned by `GetFwdXdlopsNHWCConfigList`. For example you can use aggregate like this (pseudo code):

```C++
struct GeneratedData
{
   Datatype datatype;
   Layout layout;
   Direction direction;
   PerformanceConfigAsmImplicitGemmGTCFwdXdlopsNHWC perfConfig;
}
...
static const inline std::vector<GeneratedData>&
GetFwdXdlopsNHWCConfigList()
{
    static const  std::vector<GeneratedData> kernel_param_list {
        {"fwd", "nhwc", miopenFloat,  { ... } }, // a pair of curly braces inserted
        ...
}
```

Or you can initialize a pair of structs. Or something else.

This is just a refactoring, i.e. linear transformation of the code, it I do not see any exceptional complexity. If you need assistance, please let me know. I can do that for you in one of the files.

But we definitely need to refine our data structures.

## Do we need have many members in PerformanceConfigAsmImplicitGemmGTC?

As far as I see, for PerformanceConfigAsmImplicitGemmGTCFwdXdlopsNHWC we only need to store table index and `gemm_k_global_split`. The rest of data can be read from the table right in `GetSolution()`. Please look into `SetNextValue()` and you'll see that only index and `gemm_k_global_split` are modified.

[the remark this answers is not in the document]
Comment (atamazov) [reply src/include/miopen/solver.hpp]: @junliume Actually this is not resolved.

Please let me [Resolve] "my" review comments myself, thanks.

File: src/solver/conv_asm_implicit_gemm_gtc_wrw_nhwc.cpp [hunk 2]
```diff
@@ -534,7 +670,7 @@ bool PerformanceConfigAsmImplicitGemmGTCWrwXdlopsNHWC::IsValid(const Convolution
     const auto dilation_w = ctx.kernel_size_w > 1 ? ctx.kernel_dilation_w : 1;
     const auto& pad_h     = ctx.pad_h;
     const auto& pad_w     = ctx.pad_w;
-    const auto& precision = ctx.IsFp16() ? miopenHalf : miopenFloat;
+    const auto& precision = ctx.IsFp16() ? miopenHalf : (ctx.IsBfp16() ? miopenBFloat16 : miopenFloat);
```
[the remark this answers is not in the document]
Comment (atamazov) [reply src/solver/conv_asm_implicit_gemm_gtc_wrw_nhwc.cpp:727]: Resolved at https://github.com/ROCmSoftwarePlatform/MIOpen/pull/1248/files?diff=split&w=1#r745034161

File: src/include/miopen/solver.hpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (atamazov) [reply src/include/miopen/solver.hpp]: Let's continue in #1268

File: src/solver/conv_asm_implicit_gemm_gtc_wrw_nhwc.cpp [hunk 3]
```diff
@@ -432,7 +559,11 @@ void PerformanceConfigAsmImplicitGemmGTCWrwXdlopsNHWC::HeuristicInit(const Convo
 
         CopyParameters(config_list[selected_index]);
         if(need_k_split)
+        {
+            if(ctx.IsBfp16() && tensor_b_thread_lengths[3] == 1)
+                vector_store = 1;
             gemm_k_global_split = occupancy;
+        }
```
[the remark this answers is not in the document]
Comment (atamazov) [reply src/solver/conv_asm_implicit_gemm_gtc_wrw_nhwc.cpp]: Resolved at https://github.com/ROCmSoftwarePlatform/MIOpen/pull/1248/files#r745048849

File: src/solver/conv_asm_implicit_gemm_gtc_fwd_nhwc.cpp [hunk 4]
```diff
@@ -606,6 +752,9 @@ ConvSolution ConvAsmImplicitGemmGTCDynamicFwdXdlopsNHWC::GetSolution(
     std::tie(kernel_name, block_size, grid_size) =
         GetImplicitGemmGtcDynamicFwdXdlopsNHWCKernel(ctx, config);
 
+    const auto required_workspace_size = GetWorkspaceSize(ctx);
+    result.workspce_sz                 = required_workspace_size;
```
[the remark this answers is not in the document]
Comment (shaojiewang) [reply src/solver/conv_asm_implicit_gemm_gtc_fwd_nhwc.cpp]: @atamazov variable ```required_workspace_size ``` is also used in invoker in line 965. So I think it should be kept.

File: src/include/miopen/solver.hpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (carlushuang) [reply src/include/miopen/solver.hpp]: @atamazov make sense. I'll take to shaojie about this

[the remark this answers is not in the document]
Comment (atamazov) [reply src/include/miopen/solver.hpp]: Okay.

File: src/solver/conv_asm_implicit_gemm_gtc_fwd_nhwc.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (atamazov) [reply src/solver/conv_asm_implicit_gemm_gtc_fwd_nhwc.cpp]: Ok

**Pull request merged**